modules/lang-painless/src/test/java/org/elasticsearch/painless/EqualsTests.java

```diff
@@ -133,14 +132,23 @@ public void testBranchEquals() {
         assertEquals(0, exec("def a = 1; Object b = new HashMap(); if (a === (Object)b) return 1; else return 0;"));
     }
     
-    public void testBranchEqualsDefAndPrimitive() {
+    public void testEqualsDefAndPrimitive() {
         /* This test needs an Integer that isn't cached by Integer.valueOf so we draw one randomly. We can't use any fixed integer because
          * we can never be sure that the JVM hasn't configured itself to cache that Integer. It is sneaky like that. */
         int uncachedAutoboxedInt = randomValueOtherThanMany(i -> Integer.valueOf(i) == Integer.valueOf(i), ESTestCase::randomInt);
         assertEquals(true, exec("def x = params.i; int y = params.i; return x == y;", singletonMap("i", uncachedAutoboxedInt), true));
         assertEquals(false, exec("def x = params.i; int y = params.i; return x === y;", singletonMap("i", uncachedAutoboxedInt), true));
         assertEquals(true, exec("def x = params.i; int y = params.i; return y == x;", singletonMap("i", uncachedAutoboxedInt), true));
         assertEquals(false, exec("def x = params.i; int y = params.i; return y === x;", singletonMap("i", uncachedAutoboxedInt), true));
+
+        /* Now check that we use valueOf with the boxing used for comparing primitives to def. For this we need an
+         * integer that is cached by Integer.valueOf. The JLS says 0 should always be cached. */
+        int cachedAutoboxedInt = 0;
+        assertEquals(Integer.valueOf(cachedAutoboxedInt), Integer.valueOf(cachedAutoboxedInt));
```

Comment:
I took a look because it looked fun, sorry if I'm getting things wrong. If I understand this test right, then these two values should be referentialy equal, so shouldn't this be `assertSame`?